### ISSUE TITLE
Secure Client WebSocket Connections Fail Due to `wrong version number` Errors

### ISSUE THREAD

**Issue by ITenthusiasm** [websockets/ws#2105]
### Is there an existing issue for this?

- [X] I've searched for any related issues and avoided creating a duplicate issue.

### Description

I have a Node.js server that only supports HTTPS. It creates its server using `https`, with the `cert` and `key` being generated by `mkcert`. Chrome is able to trust this certificate just fine. And thanks to `NODE_EXTRA_CA_CERTS`, Node's `https.Agent` is able to trust the certificate as well.

This https server is supplied to the `WebSocketServer` constructor. However, when I try to create a `WebSocket` client via `new WebSocket("wss://localhost:3000")`, the connection fails. Instead, I get this error:

```
Error: write EPROTO 4444288512:error:1408F10B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:332:
```

Supplying `8` or `13` to `WebSocket.protocolVersion` doesn't seem to change the outcome. And I get an error message saying I can only use `8` or `13` if I try anything else. Not sure what's going on here.

Happy to provide any additional details as needed.

---

(Note: I cannot connect to the WebSocketServer from the Browser either.)

(Note: I've already seen #1839. I can confirm that I can make HTTPS requests to my own server from Node.js, so that is not the issue. The resolution proposed in #1882 is also insufficient.)

### ws version

`7.4.5`

### Node.js Version

`16.17.0`

### System

```
  System:
    OS: macOS 12.6
    CPU: (8) x64 Intel(R) Core(TM) i7-1068NG7 CPU @ 2.30GHz
    Memory: 87.46 MB / 32.00 GB
    Shell: 5.8.1 - /bin/zsh
```

### Expected result

I would expect the connection to go through just fine (whether via Browser or via the `WebSocket` client) since both the Browser and Node.js trust my certificate.

### Actual result

Neither the browser nor the package's `WebSocket` client are able to connect over `wss`.

**Comment by lpinca** [comment]
This is is not a `ws` issue. The WebSocket protocol version has nothing to do with this. This is a TLS issue. I don't know how `mkcert` works but I guess the certificate is not trusted. Try use the [`ca`]( https://nodejs.org/api/tls.html#tlscreatesecurecontextoptions) or [`rejectUnauthorized`](https://nodejs.org/api/tls.html#tlsconnectoptions-callback) options.

**Comment by ITenthusiasm** [comment]
Is there a reason why node and the browser would trust my certificate with HTTPS but not with a web socket?

**Comment by ITenthusiasm** [comment]
I will try when I get back to my desktop

**Comment by lpinca** [comment]
> Is there a reason why node and the browser would trust my certificate with HTTPS but not with a web socket?

Does the Node.js `https` module works without the `NODE_EXTRA_CA_CERTS` env variable? Also try to connect with the plain `tls` module. This is what `ws` does by defautl.

**Comment by ITenthusiasm** [comment]
> Does the Node.js https module works without the NODE_EXTRA_CA_CERTS env variable?

Not without aggressive configuration. Without the env variable, the certificate is not trusted unless I force something like `https.Agent.globalOptions.ca` to point to the appropriate local root certificate.

Moreover, the env variable cannot be set at runtime (so even dotenv doesn't work). It has to be `EXPORT`ed by something like `.bashrc` in whatever environment Node.js is run.

---

Note that this is only for Node.js. My browser trusts the certificate regardless.

---

Will try the `rejectUnauthorized` and `tls` soon.

**Comment by ITenthusiasm** [comment]
> This is is not a ws issue. The WebSocket protocol version has nothing to do with this. This is a TLS issue. I don't know how mkcert works but I guess the certificate is not trusted. Try use the [ca](https://nodejs.org/api/tls.html#tlscreatesecurecontextoptions) or [rejectUnauthorized](https://nodejs.org/api/tls.html#tlsconnectoptions-callback) options.

Using `rejectUnauthorized: false` on the `https` server _does not_ work.

**Comment by ITenthusiasm** [comment]
@lpinca Nevermind. This is not a problem with my setup, but it's a problem with framework that I'm using. :\ It's dev server instantiates `WebSocket.Server` as is -- without any regard to the physical HTTP/HTTPS server than the actual developer is using.

In other words, my HTTPS server is encrypted, but the hot reloading dev server that Remix uses is not. I need to reach out and see if they're willing to fix this. Otherwise, there are some hacks I could do.

It seems like they're calling `WebSocket.Server({ port })`. Is there a way to supply a certificate when a `port` is specified without a `server`?

**Comment by ITenthusiasm** [comment]
From what I've heard, the Remix team is actually pretty stubborn when it comes to the dev server.

If we could get something like `tlsOptions` or `certificate`... or get two new props: `key` and `cert` for when `WebSocket.Server` is called _specifically_ with a `port`, I think it could go a really long way in general! But it would also hopefully help me convince the team to support TLS for the websocket server since it would be a simple change.

**Comment by lpinca** [comment]
> Is there a way to supply a certificate when a `port` is specified without a server.

No, the `port` option starts an internal plain HTTP server.

> If we could get something like `tlsOptions` or `certificate`... or get two new props: `key` and `cert` for when `WebSocket.Server` is called _specifically_ with a `port`, I think it could go a really long way in general! But it would also hopefully help me convince the team to support TLS for the websocket server since it would be a simple change.

It does not change much because those options would need to be forwarded by Remix to the `WebSocketServer`. In that case they could just use those options to create their own HTTPS server and pass it to the `WebSocketServer` constructor via the `server` option.

**Comment by ITenthusiasm** [comment]
> No, the port option starts an internal plain HTTP server.

Yes, I saw. Would it not be sufficient to conditionally switch between the `http` server and the `https` server depending on whether or not the certificates were added? What is this downside to this, if you don't mind my asking?

> It does not change much because those options would need to be forwarded by Remix to the WebSocketServer

Yes, I am aware. The option(s) would be applied to `remix.config.js`. Their `WebSocketServer` has access to the options exported from that file, so it would just use those. Fairly straightforward.

**Comment by lpinca** [comment]
> Would it not be sufficient to conditionally switch between the `http` server and the `https` server depending on whether or not the certificates were added? What is this downside to this, if you don't mind my asking?

Supporting all the TLS options. Because once you add support for one people will start asking for all the others, one by one.

> Yes, I am aware. The option(s) would be applied to `remix.config.js`

They still have to change this

https://github.com/remix-run/remix/blob/7b330a8beb79be6d693cfed59b436a93ad55af2e/packages/remix-dev/devServer/liveReload.ts#L18

so again, they can just create their own HTTPS server and pass it to the `WebSocketServer` constructor via the `server` option.

**Comment by ITenthusiasm** [comment]
Yeah, both of those things are valid. That makes sense. Well thanks for taking the time to engage this Issue! 
Sorry it turned out to be a dud. :confused: Hopefully you don't get these TLS queries too often.

Thanks for all the help!

**Comment by lpinca** [comment]
For feature parity having the ability to start an internal HTTPS server makes sense.  As written above an `internalHTTPServerOptions` option could be added and used for both `http.createServer()` and `https.createServer()` but that is not sufficient because there should also be a way to determine whether to use `http.createServer()` or `https.createServer()`. Another option like `useHTTPS` is required because an empty object is also valid for `https.createServer()` so we can't use a special property in the options object for this.

Overall I'm not very happy with this when the `server` option exists, but  I'm open to it.

**Comment by lpinca** [comment]
> Another option like `useHTTPS` is required because an empty object is also valid for `https.createServer()` so we can't use a special property in the options object for this.

Thinking again about this, we can check if at least one of the `pfx`, `key`, or `pskCallback` options is set and if so, use `https.createServer()` because it does not make much sense to create an `https.Server` without any of these or an `http.Server` server with any of these.

Anyway there are events like `'keylog'`, `'newSession'`, `'OCSPRequest'`, `'tlsClientError'`, etc. emitted by an `https.Server` server that should be reemitted by the `WebSocketServer` otherwise the user cannot listen to them because the `https.Server` server is created internally, so I'm still not convinced.

FWIW the same argument is also valid for the internally created `http.Server` when the `port` option is specified. Some events like `'clientError'` are currently not reemitted.

To summarize, the `server` and `noServer` options are preferable as they offer more control. The `port` option is just a quick way to start a server without customization and boilerplate code.

**Comment by ITenthusiasm** [comment]
@lpinca Understood! Thanks for sharing your additional thoughts on this!

Yeah it sounds like there might be more effort than what it's worth. Currently this issue is an edge case surrounding how Remix handles its dev server and hot module reloading. Maybe the story would be different or worth revisiting if a lot of people ran into valid use cases that required `port` + `https` _without_ an explicitly defined `server`? But that doesn't seem to be the case now. So doesn't seem to be worth tackling.
